# Evolutionary-flexibility-in-flooding-response-2019
Code and supplementary files used in analysis for "Evolutionary flexibility in flooding response circuitry in angiosperms" (2019)

**Title:** Evolutionary flexibility in flooding response circuitry in angiosperms

**Authors:**  [Mauricio A. Reynoso](https://github.com/reynosoma), [Kaisa Kajala](https://github.com/kaisakajala), [Marko Bajic](https://github.com/DealLab), Donnelly A. West, Germain Pauluzzi, Andrew Yao, Katie Hatch, Kristina  Zumstein, Margaret Woodhouse, [Joel Rodriguez-Medina](https://github.com/rodriguezmDNA), Neelima Sinha, Siobhan M. Brady, [Roger B. Deal](https://github.com/DealLab), [Julia Bailey-Serres](https://github.com/jbserres)

**Journal:** [Science](https://science.sciencemag.org/cgi/doi/10.1126/science.aax8862)

**Abstract:** 

> Flooding due to extreme weather threatens crops and ecosystems. To understand variation in gene regulatory networks activated by submergence,we conducted a high-resolution analysis of chromatin accessibility and gene expression at three scales of transcript control in four angiosperms, ranging from a dryland-adapted wild species to a wetland crop. The data define a cohort of conserved submergence-activated genes with signatures of overlapping cis-regulation by four transcription factor families. Syntenic genes are more highly expressed than nonsyntenic genes, yet both can have the cismotifs and chromatin accessibility associated with submergence up-regulation. Whereas the flexible circuitry spans the eudicot-monocot divide, the frequency of specific cis-motifs, extent of chromatin accessibility, and the degree of submergence activation is more prevalent in thewetland crop andmay have adaptive importance.



## Code used by topic

- [ATAC-seq](https://github.com/plant-plasticity/Evolutionary-flexibility-in-flooding-response-2019/tree/master/ATAC-seq)
- [RNA mapping for rice](https://github.com/plant-plasticity/Evolutionary-flexibility-in-flooding-response-2019/tree/master/Mapping-RNA-seq-Os)
- [Differential Gene Expression by limma-voom](https://github.com/plant-plasticity/Evolutionary-flexibility-in-flooding-response-2019/tree/master/DEG-analysis-limma-voom)
- [Ribotaper for periodicity](https://github.com/plant-plasticity/Evolutionary-flexibility-in-flooding-response-2019/tree/master/Ribotaper-periodicity)
- [Clustering](https://github.com/plant-plasticity/Evolutionary-flexibility-in-flooding-response-2019/blob/master/DEG-analysis-limma-voom/Scripts/Clustering_Heatmap.R)
- [Cross-species comparisons using Phytozome gene families](https://github.com/plant-plasticity/Evolutionary-flexibility-in-flooding-response-2019/tree/master/Gene-families-Phytozome)


## Other resources by topic
- [Genome browser - CoGe viewer of ATAC-seq, nuclear RNA, total RNA, TRAP RNA and Ribo-seq](https://github.com/plant-plasticity/Evolutionary-flexibility-in-flooding-response-2019/tree/master/CoGe)
- [Syntenic relationships for cross-species comparisons](https://github.com/plant-plasticity/Evolutionary-flexibility-in-flooding-response-2019/tree/master/Synteny)
- [Cytoscape files for Figure 4](https://github.com/plant-plasticity/Evolutionary-flexibility-in-flooding-response-2019/tree/master/Cytoscape)

